core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -540,11 +544,26 @@ private void updateCacheStream(long blockId) throws IOException {
     }
 
     try {
-      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-          mBlockStore.getWorkerInfoList(), getBlockSizeAllocation(mPos));
-      // If we reach here, we need to cache.
-      mCurrentCacheStream =
-          mBlockStore.getOutStream(blockId, getBlockSize(mPos), address, mOutStreamOptions);
+      List<BlockWorkerInfo> workers = mBlockStore.getWorkerInfoList();
+      // If we are reading a block from a remote worker, we shouldn't cache back to it.
+      if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
+        InetSocketAddress address = ((RemoteBlockInStream) mCurrentBlockInStream).location();
+        Iterator<BlockWorkerInfo> it = workers.iterator();
+        while (it.hasNext()) {
+          BlockWorkerInfo worker = it.next();
+          if (worker.getNetAddress().getHost().equals(address.getHostString())
+              && worker.getNetAddress().getDataPort() == address.getPort()) {
+            it.remove();
+          }
+        }
+      }
+      if (!workers.isEmpty()) {
+        WorkerNetAddress address =
```

Comment:
I think the original intention is to only cache the remote block locally, if this is the case, the dead lock won't appear.